### PR TITLE
OpenQASM local simulator

### PR DESCRIPTION
**Pull request by ajberdy**
*Issue #, if available:*

*Description of changes:*

Two simulators for simulating OpenQASM locally with Braket
- State vector simulator that isolates quantum simulation for high performance with high shot count
- Native simulator that gives user greater flexibility in specifying input/output for OpenQASM programs

Generally a WIP, but this iteration is viable enough to be merged into a feature branch

Related PRs: https://github.com/aws/amazon-braket-sdk-python/pull/354, https://github.com/aws/amazon-braket-schemas-python/pull/110

*Testing done:*

Tox, manual

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/amazon-braket-default-simulator-python/blob/main/CONTRIBUTING.md) doc
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/amazon-braket-default-simulator-python/blob/main/CONTRIBUTING.md#commit-your-change)
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/amazon-braket-default-simulator-python/blob/main/README.md) and [API docs](https://github.com/aws/amazon-braket-default-simulator-python/blob/main/CONTRIBUTING.md#documentation-guidelines) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .github/workflows/dependent-tests.yml [hunk 1]
```diff
@@ -26,9 +26,10 @@ jobs:
       with:
         python-version: ${{ matrix.python-version }}
     - name: Install dependencies
+      # TODO: change target before merging into public
       run: |
         pip install --upgrade pip
-        pip install --upgrade git+https://github.com/aws/amazon-braket-schemas-python@main
+        pip install --upgrade git+https://github.com/aws/amazon-braket-schemas-python@feature/openqasm-local-simulator
```
Comment (krneta) [inline .github/workflows/dependent-tests.yml:32]: Do you want to merge this to main?

Comment (ajberdy) [reply .github/workflows/dependent-tests.yml:32]: This is the PR into the feature branch. The line of code highlighted here points the dependent tests to the schema feature branch, which is where the schema changes will stay until we're ready to hard launch

File: .github/workflows/python-package.yml [hunk 2]
```diff
@@ -5,17 +5,19 @@ name: Python package
...
 
 jobs:
   build:
 
     runs-on: ${{ matrix.os }}
     strategy:
       matrix:
-        os: [ubuntu-latest, macos-latest, windows-latest]
+        # os: [ubuntu-latest, macos-latest, windows-latest]
+        # disabling windows for now
```
[the remark this answers is not in the document]
Comment (ajberdy) [reply .github/workflows/python-package.yml:19]: The antlr install step would have to be modified to build on windows. Since we're not planning on sticking with the antlr parser, that would be wasted work. Todo: reenable windows build once we remove the antlr dependency

File: lint-openqasm.sh [hunk 3]
```diff
@@ -0,0 +1,17 @@
+isort src/braket/default_simulator/openqasm \
```
[the remark this answers is not in the document]
Comment (ajberdy) [reply lint-openqasm.sh]: artifact of early development, deleted

File: setup.py [hunk 4]
```diff
@@ -19,24 +24,123 @@
 with open("src/braket/default_simulator/_version.py") as f:
     version = f.readlines()[-1].split()[-1].strip("\"'")
 
+
+class InstallOQ3Command(distutils.cmd.Command):
```
[the remark this answers is not in the document]
Comment (ajberdy) [reply setup.py:28]: This is an automation of the install process for the antlr openqasm parser. This is a stop-gap until we can swap parsers.

File: src/braket/default_simulator/linalg_utils.py [hunk 5]
```diff
@@ -119,3 +119,15 @@ def _get_target_permutation(targets: Sequence[int]) -> Sequence[int]:
     return np.ravel_multi_index(
         basis_states[:, np.argsort(np.argsort(targets))].T, [2] * len(targets)
     )
+
+
+def controlled_unitary(unitary, neg=False):
```
[the remark this answers is not in the document]
Comment (ajberdy) [reply src/braket/default_simulator/linalg_utils.py:124]: Ah, yes I'll do another pass to get all of those that I've missed. Good call

[the remark this answers is not in the document]
Comment (krneta) [reply src/braket/default_simulator/linalg_utils.py:124]: You could also try using our codestyle linter to see if it comes up with something useful.

[the remark this answers is not in the document]
Comment (ajberdy) [reply src/braket/default_simulator/linalg_utils.py:124]: Our set up linters for this repo don't complain. I think we might have to move forward with pylint to get better/more opinionated linting

File: src/braket/default_simulator/openqasm/data_manipulation.py [hunk 6]
```diff
@@ -0,0 +1,656 @@
...
+    raise TypeError(f"Cannot cast {type(variable).__name__} into {into.__name__}.")
+
+
+@cast_to.register
+def _(into: BitType, variable: Union[BooleanLiteral, ArrayLiteral, BitstringLiteral]):
+    """
+    Bit types can be sized or not, represented as Boolean literals or Array literals.
+    Sized bit types can be instantiated with a Bitstring literal or Array literal.
+    """
+    if not into.size:
```
Comment (krneta) [inline src/braket/default_simulator/openqasm/data_manipulation.py:256]: Is this the same as checking if isinstance(variable, BooleanLiteral), or not?

File: src/braket/default_simulator/simulator.py [hunk 7]
```diff
@@ -135,45 +86,21 @@ def _validate_ir_results_compatibility(self, circuit_ir):
...
-        if no_noise and _NOISE_INSTRUCTIONS.intersection(supported_instructions):
-            warnings.warn(
-                'You are running a noise-free circuit on the density matrix simulator. \
-Consider running this circuit on the state vector simulator: LocalSimulator("default") \
-for a better user experience.'
-            )
-
     @staticmethod
-    def _validate_shots_and_ir_results(shots: int, circuit_ir: Program, qubit_count: int) -> None:
+    def _validate_shots_and_ir_results(shots: int, results, qubit_count: int) -> None:
```
Comment (krneta) [inline src/braket/default_simulator/simulator.py:90]: type.

File: src/braket/default_simulator/openqasm/data_manipulation.py (hunk 6 above)
Comment (ajberdy) [reply src/braket/default_simulator/openqasm/data_manipulation.py:256]: You are correct in that that is when that branch happens. The slight difference would be in error handling. I think the logic is a little cleaner if split by intent than by type of thing being cast. But yes with perfect code, those should be equivalent

File: src/braket/default_simulator/openqasm/interpreter.py [hunk 8]
```diff
@@ -0,0 +1,600 @@
...
+                # statement is a quantum phase instruction
+                else:
+                    inlined_body.append(statement)
+            # this includes converted phase instructions
+            if isinstance(statement, QuantumGate):
+                gate_name = statement.name.name
+                statement.arguments = self.visit(statement.arguments)
+                statement.modifiers = self.visit(statement.modifiers)
+                statement.qubits = self.visit(statement.qubits)
+                if gate_name == "U":
```
[the remark this answers is not in the document]
Comment (ajberdy) [reply src/braket/default_simulator/openqasm/interpreter.py:303]: `U` is the name of the singular builtin native gate in the OpenQASM language specification. For that reason, this branch is the base case of this recursive function for inlining the gate definition body

File: src/braket/default_simulator/openqasm/braket_result_pragmas.py [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+    Variance,
+)
+from openqasm3.parser import parse
+
+from .dist.BraketPragmasLexer import BraketPragmasLexer
+from .dist.BraketPragmasParser import BraketPragmasParser
+from .dist.BraketPragmasVisitor import BraketPragmasVisitor
+
+
+class BraketPragmaNodeVisitor(BraketPragmasVisitor):
```
Comment (speller26) [inline src/braket/default_simulator/openqasm/braket_result_pragmas.py:18]: Not a big deal, but this looks like the _perfect_ place for `singledispatch`.

File: src/braket/default_simulator/linalg_utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (speller26) [reply src/braket/default_simulator/linalg_utils.py:124]: Not sure if I've asked before, but what is the `neg` flag for? It doesn't swap the qubits, and neither does it invert the controlled unitary.

File: src/braket/default_simulator/openqasm/interpreter.py [hunk 10]
```diff
@@ -0,0 +1,595 @@
...
+        program = parse(source)
+        self.visit(program)
+        return self.context.circuit
+
+    def run(self, source, inputs=None, is_file=False):
+        """Interpret an OpenQASM program and return the program state"""
+        self.build_circuit(source, inputs, is_file)
+        return self.context
+
+    @singledispatchmethod
```
Comment (speller26) [inline src/braket/default_simulator/openqasm/interpreter.py:116]: How are we planning to handle the fact that this doesn't work with Python 3.7?

File: .github/workflows/python-package.yml [hunk 11]
```diff
@@ -5,17 +5,19 @@ name: Python package
...
+        os: [ubuntu-latest, macos-latest]
         python-version: [3.7, 3.8, 3.9]
```
Comment (speller26) [inline .github/workflows/python-package.yml:21]: Surprised `@singledispatchmethod` isn't breaking 3.7

Comment (ajberdy) [reply .github/workflows/python-package.yml:21]: I added a python 3.7 workaround for singledispatchmethod here: https://github.com/aws/amazon-braket-default-simulator-python/blob/openqasm-local-simulator/src/braket/default_simulator/openqasm/data_manipulation.py#L640

File: src/braket/default_simulator/openqasm/braket_result_pragmas.py (hunk 9 above)
Comment (ajberdy) [reply src/braket/default_simulator/openqasm/braket_result_pragmas.py:18]: Actually this is the signature automatically created and used by antlr, just following an autogenerated interface here

File: src/braket/default_simulator/openqasm/interpreter.py (hunk 10 above)
Comment (ajberdy) [reply src/braket/default_simulator/openqasm/interpreter.py:116]: Responded above, but I added an implementation [here](https://github.com/aws/amazon-braket-default-simulator-python/blob/openqasm-local-simulator/src/braket/default_simulator/openqasm/data_manipulation.py#L640)

File: src/braket/default_simulator/gate_operations.py [hunk 12]
```diff
@@ -230,7 +230,7 @@ def matrix(self) -> np.ndarray:
             1
             / np.sqrt(2)
             * np.array(
-                [[0, 1, 0, 1.0j], [1, 0, -1.0j, 0], [0, 1.0j, 0, 1], [-1.0j, 0, 1, 0]],
+                [[0, 0, 1, 1.0j], [0, 0, 1.0j, 1], [1, -1.0j, 0, 0], [-1.0j, 1, 0, 0]],
```
Comment (ajberdy) [inline src/braket/default_simulator/gate_operations.py:233]: This change to reflect the [bdk implementation](https://github.com/aws/amazon-braket-sdk-python/blob/main/src/braket/circuits/gates.py#L1324)

**Comment by ajberdy** [comment]
Tests failing due to rounding error that doesn't show up locally, will fix tomorrow

File: test/unit_tests/braket/default_simulator/openqasm/test_interpreter.py [hunk 13]
```diff
@@ -0,0 +1,1473 @@
...
+    """
+    undefined_gate = "Gate undef_y is not defined."
+    with pytest.raises(ValueError, match=undefined_gate):
+        Interpreter().run(qasm)
+
+
+def test_gate_call():
+    qasm = """
+    float[128] my_pi = π;
+    gate x a { U(π, 0, my_pi) a; }
```
Comment (krneta) [inline test/unit_tests/braket/default_simulator/openqasm/test_interpreter.py:662]: Can we test with some non-zero, non-pi values (ie. the first last parameter be varied)?

File: test/unit_tests/braket/default_simulator/openqasm/test_interpreter.py [hunk 14]
```diff
@@ -0,0 +1,1473 @@
...
+    gate x2(p) a { U(π, 0, p) a; }
+
+    qubit q1;
+    qubit q2;
+    qubit[2] qs;
+
+    U(π, 0, my_pi) q1;
+    x q2;
+    x2(my_pi) qs[0];
```
Comment (krneta) [inline test/unit_tests/braket/default_simulator/openqasm/test_interpreter.py:671]: can we make this qs[1]?

File: test/unit_tests/braket/default_simulator/openqasm/test_interpreter.py [hunk 15]
```diff
@@ -0,0 +1,1473 @@
...
+    s q1;   // again
+    inv @ z q1; // inv z
+    """
+    circuit = Interpreter().build_circuit(qasm)
+    simulation = StateVectorSimulation(5, 1, 1)
+    simulation.evolve(circuit.instructions)
+    assert np.allclose(simulation.state_vector, np.array([0] * 31 + [1]))
+
+
+def test_measurement():
```
Comment (krneta) [inline test/unit_tests/braket/default_simulator/openqasm/test_interpreter.py:892]: what is this testing?

File: test/unit_tests/braket/default_simulator/openqasm/test_interpreter.py [hunk 16]
```diff
@@ -0,0 +1,1473 @@
...
+    context = Interpreter().run(qasm)
+    assert convert_bool_array_to_string(context.get_value("and")) == "0100"
+    assert convert_bool_array_to_string(context.get_value("or")) == "1101"
+    assert convert_bool_array_to_string(context.get_value("xor")) == "1001"
+    assert convert_bool_array_to_string(context.get_value("lshift")) == "0100"
+    assert convert_bool_array_to_string(context.get_value("rshift")) == "0011"
+    assert convert_bool_array_to_string(context.get_value("flip")) == "1010"
+    assert context.get_value("gt") == BooleanLiteral(False)
+    assert context.get_value("lt") == BooleanLiteral(True)
+    assert context.get_value("ge") == BooleanLiteral(False)
```
Comment (krneta) [inline test/unit_tests/braket/default_simulator/openqasm/test_interpreter.py:1093]: doesn't really test ge is any different from gt and le any different from lt

File: test/unit_tests/braket/default_simulator/test_openqasm_state_vector_simulator.py [hunk 17]
```diff
@@ -0,0 +1,266 @@
...
+
+    #pragma braket result amplitude '00', '01', '10', '11'
+    """
+    simulator = OpenQASMStateVectorSimulator()
+    result = simulator.run(Program(source=qasm))
+    sv = [result.resultTypes[0].value[state] for state in ("00", "01", "10", "11")]
+    assert np.allclose(sv, [-1 / np.sqrt(2), 0, 0, 1 / np.sqrt(2)])
+
+
+@pytest.fixture
```
Comment (krneta) [inline test/unit_tests/braket/default_simulator/test_openqasm_state_vector_simulator.py:46]: can you put fixtures before tests?

File: test/unit_tests/braket/default_simulator/test_openqasm_state_vector_simulator.py [hunk 18]
```diff
@@ -0,0 +1,266 @@
...
+
+
+@pytest.mark.xfail(reason="result types not translated for shots simulation until bdk")
+def test_adder(sv_adder):
+    simulator = OpenQASMStateVectorSimulator()
+    inputs = {"a_in": 7, "b_in": 3}
+    result = simulator.run(Program(source="adder.qasm", inputs=inputs), shots=100)
+    expected_probs = np.zeros(2**5)
+    expected_probs[10] = 1
+    assert np.allclose(result.resultTypes[0].value, expected_probs)
```
Comment (krneta) [inline test/unit_tests/braket/default_simulator/test_openqasm_state_vector_simulator.py:107]: Do these actually run?

File: src/braket/default_simulator/openqasm/data_manipulation.py [hunk 19]
```diff
@@ -0,0 +1,656 @@
...
+
+def is_none_like(value: Any) -> bool:
+    """Returns whether value is None or an Array of Nones"""
+    if isinstance(value, ArrayLiteral):
+        return all(is_none_like(v) for v in value.values)
+    return value is None
+
+
+"""
+Helper functions for working with indexed values
```
[the remark this answers is not in the document]
Comment (ajberdy) [reply src/braket/default_simulator/openqasm/data_manipulation.py:360]: I can do that if you think it would be cleaner, I just went with one file for all helper methods, but open to shaking it up. Are you asking if I can move these functions to another file without creating circular dependencies?

File: src/braket/default_simulator/openqasm/data_manipulation.py [hunk 20]
```diff
@@ -0,0 +1,656 @@
...
+
+def convert_index(index: Union[RangeDefinition, IntegerLiteral]) -> Union[int, slice]:
+    """Convert unspecified index type to Python object"""
+    if isinstance(index, RangeDefinition):
+        return convert_range_def_to_slice(index)
+    else:  # IntegerLiteral:
+        return index.value
+
+
+def flatten_indices(indices: List[IndexElement]) -> list:
```
[the remark this answers is not in the document]
Comment (ajberdy) [reply src/braket/default_simulator/openqasm/data_manipulation.py:436]: Most of the functionality here is tested through the interpreter. ie instead of
```
assert flatten_indices([[1, 2], [3], [4, 5]]) == [1, 2, 3, 4, 5]
```
The test would be
```
qasm = """
array[int[8], 6, 6, 6, 6, 6] arr = {...};
int[8] get_1 = arr[1, 2][3][4, 5];
int[8] get_2 = arr[1, 2, 3, 4, 5];
int[8] get_3 = arr[1][2][3][4][5];
"""
# run program
# assert get_1 == get_2 == get_3
```

Maybe we can chat about testing strategy here, I'd love to hear your thoughts

File: src/braket/default_simulator/openqasm/data_manipulation.py [hunk 21]
```diff
@@ -0,0 +1,656 @@
...
+    return ArrayLiteral([wrap_value_into_literal(v) for v in value])
+
+
+"""
+Python 3.7 compatibility
+"""
+
+
+def singledispatchmethod(func):
+    """Implement singledispatchmethod for Python 3.7"""
```
[the remark this answers is not in the document]
Comment (ajberdy) [reply src/braket/default_simulator/openqasm/data_manipulation.py:648]: Yes. If that's problematic, we could condition the import on a python version check. Do you see a reason to do that?

File: src/braket/default_simulator/openqasm/program_context.py [hunk 22]
```diff
@@ -0,0 +1,559 @@
+from typing import Any, Dict, Hashable, Iterable, List, Optional, Tuple, Type, Union
```
[the remark this answers is not in the document]
Comment (ajberdy) [reply src/braket/default_simulator/openqasm/program_context.py:1]: The AST gives us a static representation of a program. To actually execute a program, we need to track a state; for example, variables, scope, etc. So when visiting the AST, if we instantiate a new variable, the ProgramContext object would track its value. If we enter a new scope, that is tracked in ProgramContext. If a quantum operation is called, or a result type declared, that is tracked by ProgramContext. This allows the Interpreter implementation to be stateless, except for the explicit ProgramContext object it accesses

File: src/braket/default_simulator/openqasm_state_vector_simulator.py [hunk 23]
```diff
@@ -0,0 +1,102 @@
...
+    def properties(self) -> GateModelSimulatorDeviceCapabilities:
+        """
+        Device properties for the OpenQASMStateVectorSimulator.
+
+        Returns:
+            GateModelSimulatorDeviceCapabilities: Device capabilities for this simulator.
+        """
+        observables = ["x", "y", "z", "h", "i", "hermitian"]
+        max_shots = sys.maxsize
+        qubit_count = 26
```
[the remark this answers is not in the document]
Comment (ajberdy) [reply src/braket/default_simulator/openqasm_state_vector_simulator.py:53]: That's a good question. This was already set to 26 for our existing local state vector simulators (and 13 for the dm local simulator)

File: test/unit_tests/braket/default_simulator/openqasm/test_interpreter.py [hunk 24]
```diff
@@ -0,0 +1,1473 @@
...
+    assert context.get_type("uninitialized") == IntType(IntegerLiteral(8))
+    assert context.get_type("pos") == IntType(IntegerLiteral(8))
+    assert context.get_type("neg") == IntType(IntegerLiteral(5))
+    assert context.get_type("pos_overflow") == IntType(IntegerLiteral(3))
+    assert context.get_type("neg_overflow") == IntType(IntegerLiteral(3))
+
+    assert context.get_value("uninitialized") is None
+    assert context.get_value("pos") == IntegerLiteral(10)
+    assert context.get_value("neg") == IntegerLiteral(-4)
+    assert context.get_value("pos_overflow") == IntegerLiteral(1)
```
[the remark this answers is not in the document]
Comment (ajberdy) [reply test/unit_tests/braket/default_simulator/openqasm/test_interpreter.py:79]: OpenQASM uses a fixed point representation for its numbers, with a sign bit and magnitude bits, rather than the common two's complement representation [doc source](https://openqasm.com/language/types.html#classical-types:~:text=We%20use%20the%20notation%20s%3Am%3Af%20to%20denote%20the%20width%20and%20precision%20of%20fixed%20point%20numbers%2C%20where%20s%20is%20the%20number%20of%20sign%20bits%2C%20m%20is%20the%20number%20of%20integer%20bits%2C%20and%20f%20is%20the%20number%20of%20fractional%20bits.). Overflow is not specifically described in the spec, so it is generally accepted as implementation-specific.

File: test/unit_tests/braket/default_simulator/openqasm/test_interpreter.py [hunk 25]
```diff
@@ -0,0 +1,1473 @@
...
+        (UserWarning, neg_overflow),
+    }
+
+
+def test_uint_declaration():
+    qasm = """
+    uint[8] uninitialized;
+    uint[8] pos = 10;
+    uint[3] pos_not_overflow = 5;
+    uint[3] pos_overflow = 8;
```
[the remark this answers is not in the document]
Comment (ajberdy) [reply test/unit_tests/braket/default_simulator/openqasm/test_interpreter.py:94]: Sure, I'll add that to the test

File: test/unit_tests/braket/default_simulator/openqasm/test_interpreter.py [hunk 26]
```diff
@@ -0,0 +1,1473 @@
...
+    """
+    no_reset = "Reset not supported"
+    with pytest.raises(NotImplementedError, match=no_reset):
+        Interpreter().run(qasm)
+
+
+def test_for_loop():
+    qasm = """
+    int[8] x = 0;
+    int[8] y = 0;
```
[the remark this answers is not in the document]
Comment (ajberdy) [reply test/unit_tests/braket/default_simulator/openqasm/test_interpreter.py:410]: I don't think that adds anything to this test. The purpose of this test is to make sure the for loops are working correctly.

File: test/unit_tests/braket/default_simulator/openqasm/test_interpreter.py (hunk 13 above)
Comment (ajberdy) [reply test/unit_tests/braket/default_simulator/openqasm/test_interpreter.py:662]: This test gives three equivalent ways of calling a quantum gate:
- inline using U
- with a gate static gate definition
- with a parameterized gate

This test focuses on asserting that the various identifiers are resolved to a value (gate name, variable in gate definition, variable in gate call, variable in gate parameterization). We test various values in U extensively, ie in test_braket_gates.

File: test/unit_tests/braket/default_simulator/openqasm/test_interpreter.py (hunk 15 above)
Comment (ajberdy) [reply test/unit_tests/braket/default_simulator/openqasm/test_interpreter.py:892]: We expect to support more robust outputs using the measure instruction in the future, so this will be more fleshed out then

File: test/unit_tests/braket/default_simulator/test_openqasm_state_vector_simulator.py (hunk 18 above)
Comment (ajberdy) [reply test/unit_tests/braket/default_simulator/test_openqasm_state_vector_simulator.py:107]: Yes, why? I will change this test to be a little more precise instead of xfailing though.

File: src/braket/default_simulator/openqasm/data_manipulation.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (krneta) [reply src/braket/default_simulator/openqasm/data_manipulation.py:362]: Yes. I mean I know we can deal with circular dependencies in either case, but if there is a clean split between these it makes it all the more obvious we should split them. 

These kind of files get to be unwieldily quickly and if they (at least) have a name that corresponds to the functionality it's a warning sign when you start adding functions that don't fit the name (kind of similar to classes and putting non-class-specific stuff in the class).

File: src/braket/default_simulator/openqasm/data_manipulation.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (krneta) [reply src/braket/default_simulator/openqasm/data_manipulation.py:447]: Yeah, but then I wonder why it's not just a private function in the interpreter? 

It's a little easier to see the test coverage when you're testing the function directly. Ie. 
Are we testing all the boundary cases ([], [1, [2, 3]][4][][5][None][6]), etc

File: src/braket/default_simulator/openqasm/data_manipulation.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (krneta) [reply src/braket/default_simulator/openqasm/data_manipulation.py:685]: Not really. I think it's fine to do this. I just wonder if there's some way to make it easy on ourselves to recognize this should be removed when we stop supporting 3.7 (which is getting deprecated pretty soon).

File: src/braket/default_simulator/openqasm/program_context.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (krneta) [reply src/braket/default_simulator/openqasm/program_context.py:1]: I'm with you, and I think that's the right approach. The state of parsing the AST should be in the context (or, at least, outside of the AST itself).

File: src/braket/default_simulator/simulator.py [hunk 27]
```diff
@@ -215,6 +185,46 @@ def _validate_result_types_qubits_exist(
...
+            TypeError: If any the specified result types are not supported
+        """
+        circuit_instruction_names = [
+            instr.__class__.__name__.lower().replace("_", "") for instr in circuit_ir.instructions
+        ]
+        supported_instructions = frozenset(
+            op.lower().replace("_", "")
+            for op in self.properties.action[self.device_action_type].supportedOperations
+        )
+        no_noise = True
```
[the remark this answers is not in the document]
Comment (krneta) [reply src/braket/default_simulator/simulator.py:207]: I think it's fine to address it in a future PR, but it likely doesn't belong here.

File: test/unit_tests/braket/default_simulator/openqasm/test_interpreter.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (krneta) [reply test/unit_tests/braket/default_simulator/openqasm/test_interpreter.py:419]: Your loop could be overwriting the initialization value and you wouldn't know it. Fine for whitebox testing though, I suppose.

File: test/unit_tests/braket/default_simulator/openqasm/test_interpreter.py (hunk 13 above)
Comment (krneta) [reply test/unit_tests/braket/default_simulator/openqasm/test_interpreter.py:712]: That's the problem with such general tests. Yes, you're testing what you're saying. What you're not testing is that when you initialize a quantum gate it can be initialized to anything except for U(π, 0, π)

File: test/unit_tests/braket/default_simulator/openqasm/test_interpreter.py (hunk 16 above)
Comment (krneta) [reply test/unit_tests/braket/default_simulator/openqasm/test_interpreter.py:1143]: Sure, it doesn't prevent it from being changed in the wrong way, but yes, I'm being too pedantic here.

File: src/braket/default_simulator/openqasm/data_manipulation.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (speller26) [reply src/braket/default_simulator/openqasm/data_manipulation.py:447]: Yeah, I don't see any reason to just have it private in program_context; in general, we should make things private as much as we can, and only the core API public. This applies to many of the utility methods in this file.

File: src/braket/default_simulator/state_vector_simulation.py [hunk 28]
```diff
@@ -131,16 +137,28 @@ def density_matrix(self) -> np.ndarray:
...
     def state_with_observables(self) -> np.ndarray:
         """
-        np.ndarray: The state vector diagonalized in the basis of the measured observables.
+        State vector with observables applied.
 
         Raises:
             RuntimeError: If observables have not been applied
+
+        Returns:
+            np.ndarray: The state vector diagonalized in the basis of the measured observables.
```
Comment (speller26) [inline src/braket/default_simulator/state_vector_simulation.py:146]: IIRC, properties are actually documented the previous way

File: src/braket/default_simulator/state_vector_simulation.py [hunk 29]
```diff
@@ -66,6 +66,12 @@ def __init__(self, qubit_count: int, shots: int, batch_size: int):
         self._post_observables = None
 
     def evolve(self, operations: List[GateOperation]) -> None:
+        """
+        Evolve simulation by applying the provided operations.
+
+        Args:
+            operations (List[GateOperation]): Operations to apply.
+        """
```
Comment (speller26) [inline src/braket/default_simulator/state_vector_simulation.py:74]: Are this and `expectation` docstring really necessary, or does this inherit documentation from a superclass? If the former, nice catch and thanks for adding!

File: src/braket/default_simulator/state_vector_simulator.py [hunk 30]
```diff
@@ -18,21 +18,36 @@
...
     def initialize_simulation(self, **kwargs) -> StateVectorSimulation:
+        """
+        Initialize state vector simulation.
+
+        Kwargs:
+            qubit_count (int), shots (int), batch_size (int)
+
+        Returns:
+            StateVectorSimulation: Initialized simulation.
+        """
```
Comment (speller26) [inline src/braket/default_simulator/state_vector_simulator.py:37]: Do these inherit documentation?

File: src/braket/default_simulator/simulator.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (speller26) [reply src/braket/default_simulator/simulator.py:207]: This variable exists for the _sole purpose_ of telling the user that they're running a noiseless circuit on a noisy simulator, which is really suboptimal

File: setup.cfg [hunk 31]
```diff
@@ -25,3 +25,8 @@ exclude =
     bin
     build
     venv
+    src/braket/default_simulator/openqasm/dist
+    openqasm
+    local-braket
+    amazon-braket-schemas-python
+    demo
```
[the remark this answers is not in the document]
Comment (ajberdy) [reply setup.cfg:32]: I added some of these to clean up linting output in my own workspace, I'll remove 3/4 of these lines before merging

File: src/braket/default_simulator/openqasm/circuit.py [hunk 32]
```diff
@@ -0,0 +1,60 @@
+from __future__ import annotations
+
+from typing import List, Optional
+
+from braket.ir.jaqcd.program_v1 import Results
```
[the remark this answers is not in the document]
Comment (ajberdy) [reply src/braket/default_simulator/openqasm/circuit.py:5]: This is a union defined in that file:
```
Results = Union[Amplitude, Expectation, Probability, Sample, StateVector, DensityMatrix, Variance]
```
I would move it, but that could be a breaking change.

Our results.py file is also in ir.jaqcd, so I expect we'll be doing an overhaul of this schema structure in the near future anyways.

File: src/braket/default_simulator/state_vector_simulation.py (hunk 29 above)
Comment (ajberdy) [reply src/braket/default_simulator/state_vector_simulation.py:74]: Ah good call, didn't realize this

File: src/braket/default_simulator/state_vector_simulation.py (hunk 28 above)
Comment (ajberdy) [reply src/braket/default_simulator/state_vector_simulation.py:146]: Good call, reverting

File: test/unit_tests/braket/default_simulator/openqasm/test_interpreter.py (hunk 13 above)
Comment (ajberdy) [reply test/unit_tests/braket/default_simulator/openqasm/test_interpreter.py:712]: I feel that this is tested extensively with the correctness test for `braket_gates.inc`

File: src/braket/default_simulator/state_vector_simulator.py (hunk 30 above)
Comment (ajberdy) [reply src/braket/default_simulator/state_vector_simulator.py:37]: Yes, but more general:
```
"""Initializes simulation with keyword arguments"""
```

**Review by krneta — APPROVED** [review]
